Add Microservices Bulkhead pattern implementation (#3228)

beingadish opened the pull request:
## What does this PR do?

- Implements resource isolation with dedicated thread pools
- Demonstrates fail-fast behavior under load

Fixes #3228

beingadish committed:
Add Microservices Bulkhead pattern implementation

- Implements resource isolation with dedicated thread pools
- Demonstrates fail-fast behavior under load
- Fixes #3228

beingadish committed:
Merge branch 'master' into feature/microservices-bulkhead

beingadish commented:
Hi @iluwatar , kindly merge this PR

iluwatar reviewed (changes requested):
Add the new module to parent pom.xml, otherwise it's not built by CI at all.

beingadish committed:
Merge branch 'iluwatar:master' into feature/microservices-bulkhead

beingadish committed:
build: add microservices-bulkhead module to root pom

beingadish committed:
build: update dependencies in microservices-bulkhead

* Remove `org.projectlombok:lombok` dependency.
* Add explicit version `3.27.7` to the `org.assertj:assertj-core` test dependency.

beingadish committed:
feat(microservices-bulkhead): refactor implementation and add unit tests

Refactor all core classes and add comprehensive test coverage for the
Microservices Bulkhead pattern.

Refactored:
- BulkheadService: overhaul thread pool isolation logic
- App: update demo to reflect revised service APIs
- BackgroundService, UserService: align with updated BulkheadService
- Task, TaskType: revise model structure and toString representation
- pom.xml: update module dependency configuration

Added tests:
- AppTest: verifies main entry point runs without exception
- BulkheadServiceTest: task acceptance, overflow rejection, active thread
  count, queue size reporting, and graceful shutdown
- BackgroundServiceTest: task submission, bulkhead overflow, and isolation
  from UserService overload
- UserServiceTest: user request acceptance and isolation from overloaded
  BackgroundService
- TaskTest: name, type, duration getters and toString format
- TaskTypeTest: enum values USER_REQUEST and BACKGROUND_PROCESSING

beingadish committed:
Refactored:
- Task: used Lombok @Getter

beingadish commented on microservices-bulkhead/src/main/java/com/iluwatar/bulkhead/Task.java:
```diff
@@ -0,0 +1,26 @@
+ * information is useful for logging, monitoring, and demonstrating the bulkhead
+ * pattern's behavior under different loads.
+ */
+@Getter
+@RequiredArgsConstructor
+public class Task {
```
Lombok is already declared as a global dependency in the root `pom.xml` (`<dependencies>` section, not just `<dependencyManagement>`), with annotation processing configured in maven-compiler-plugin. All child modules inherit it automatically; no explicit declaration needed in the module `pom.xml`.

beingadish commented on microservices-bulkhead/pom.xml:
```diff
@@ -0,0 +1,138 @@
+            <scope>runtime</scope>
+        </dependency>
+
+        <!-- JUnit Jupiter for testing -->
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+
```
`junit-jupiter-api` is intentionally absent from the root `pom.xml`. The project uses `junit-jupiter-engine` which transitively pulls in `junit-jupiter-api`. This is consistent with every other module in the repo — none of them explicitly declare `junit-jupiter-api` either.

beingadish committed:
style: add license headers and apply Google Java Format

iluwatar commented:
@beingadish consider rebasing on top of `master`. There were spotless issues that were fixed just moment ago.
